src/target/riscv/riscv-013_reg.c

```diff
@@ -13,73 +13,452 @@
 #include "debug_defines.h"
 #include <helper/time_support.h>
 
-static int riscv013_reg_get(struct reg *reg)
+/* Start of register fetch/send definitions.
+ * - "reg_fetcher" (named "fetch_*") -- loads the value from target into reg
+ *   cache entry. Modifyes only "reg->value".
+ * - "reg_sender" (named "send_*") -- stores the value from "buf" to the
+ *   target. Does not modify "reg" at all.
+ */
+typedef int (*reg_fetcher)(struct reg *reg);
+typedef int (*reg_sender)(const struct reg *reg, const uint8_t *buf);
+
+/* TODO: Introduce separate "fetch_*" and "send_*" functions for register
+ * classes (xregs, fregs, csrs).
+ */
+static int fetch_riscv013_reg(struct reg *reg)
 {
-	struct target *target = riscv_reg_impl_get_target(reg);
+	riscv_reg_t value;
+	int res = riscv013_get_register(riscv_reg_impl_get_target(reg),
+			&value, reg->number);
+	if (res != ERROR_OK)
+		return res;
+	buf_set_u64(reg->value, 0, reg->size, value);
+	return res;
+}
 
-	/* TODO: Hack to deal with gdb that thinks these registers still exist. */
-	if (reg->number > GDB_REGNO_XPR15 && reg->number <= GDB_REGNO_XPR31 &&
-			riscv_supports_extension(target, 'E')) {
-		buf_set_u64(reg->value, 0, reg->size, 0);
-		return ERROR_OK;
-	}
+static int send_riscv013_reg(const struct reg *reg, const uint8_t *buf)
+{
+	assert(reg->size <= 64);
+	const riscv_reg_t value = buf_get_u64(buf, 0, reg->size);
+	return riscv013_set_register(riscv_reg_impl_get_target(reg),
+			reg->number, value);
+}
 
-	if (reg->number >= GDB_REGNO_V0 && reg->number <= GDB_REGNO_V31) {
-		if (riscv013_get_register_buf(target, reg->value, reg->number) != ERROR_OK)
-			return ERROR_FAIL;
+/* TODO: Inline "riscv013_*_register_buf" into "vreg" accessors.
+ */
+static int fetch_vreg(struct reg *reg)
+{
+	return riscv013_get_register_buf(riscv_reg_impl_get_target(reg),
+			reg->value, reg->number);
+}
 
-		reg->valid = riscv_reg_impl_gdb_regno_cacheable(reg->number, /* is write? */ false);
-	} else {
-		uint64_t value;
-		int result = riscv_reg_get(target, &value, reg->number);
-		if (result != ERROR_OK)
-			return result;
-		buf_set_u64(reg->value, 0, reg->size, value);
+static int send_vreg(const struct reg *reg, const uint8_t *buf)
+{
+	return riscv013_set_register_buf(riscv_reg_impl_get_target(reg),
+			reg->number, reg->value);
+}
+
+/* End of register fetch/send definitions. */
+/* Start of cache entry utils. */
+
+static void set_cache_value(struct reg *reg, const uint8_t *buf)
+{
+	assert(reg);
+	assert(reg->value);
+	assert(reg->size);
```

Comment:
Could the check for an existing register be improved like this? Would it make sense to you?

```
static inline bool riscv_reg_impl_is_existing(const struct reg *reg)
{
	assert(riscv_reg_impl_is_initialized(reg));
	assert(reg->exist);
	assert(reg->size > 0);
	assert(reg->value);
}
```

```suggestion
	assert(riscv_reg_impl_is_existing(reg));
```

Reply:
Great suggestion! Thanks! Addressed